CSLabs.Api/Services/TestVmConnectionService.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using CSLabs.Api.Models;
+using CSLabs.Api.Proxmox;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+
+namespace CSLabs.Api.Services
+{
+    public class TestVmConnectionService
+    {
+        private DefaultContext Context { get; }
+
+        private ProxmoxManager ProxmoxManager;
+        
+        public TestVmConnectionService(DefaultContext context)
+        {
+            Context = context;
+        }
+        
+        public async Task<IActionResult> TestLabVmConnection()
+        {
+            var hypervisors = await Context.Hypervisors
+                .Include(h => h.HypervisorNodes)
+                .ToListAsync();
+        
+            var labVms = await Context.LabVms
+                .Include(v => v.Id)
+                .ToListAsync();
+
+            foreach (var hypervisor in hypervisors) 
+            {
+                var api = ProxmoxManager.GetProxmoxApi(hypervisor.HypervisorNodes.First());
+                
+                foreach (var labVm in labVms) 
+                { 
+                    try 
+                    { 
+                        var vmStatus = await api.GetVmStatus(labVm.Id); 
+                        if (vmStatus.IsStopped())  // vm is down
```

Comment:
You can shorten this however a user may have a vm powered down on purpose. This needs to be accounted for.